src/main/scala/edu/ie3/simona/sim/setup/ExtSimSetup.scala

```diff
@@ -0,0 +1,248 @@
+      case (setupData, connection) =>
+        connection match {
+          case extEvDataConnection: ExtEvDataConnection =>
+            if (setupData.evDataConnection.nonEmpty) {
+              throw ServiceException(
+                s"Trying to connect another EvDataConnection. Currently only one is allowed."
+              )
+            }
+
+            val serviceRef = setupInputService(
+              extEvDataConnection,
+              ExtEvDataService.props,
+              "ExtEvDataService",
+              InitExtEvData,
+            )
+
+            extSimSetupData.update(extEvDataConnection, serviceRef)
+
+          case otherConnection =>
+            log.warn(
+              s"There is currently no implementation for the connection: $otherConnection."
+            )
+            setupData
+        }
+    }
+
+    // validate data
+    validatePrimaryData(updatedSetupData.extPrimaryDataServices.keySet)
+
+    updatedSetupData
+  }
+
+  /** Method for setting up an external service, that provides input data.
+    * @param extInputDataConnection
+    *   the data connection
+    * @param props
+    *   function to create the service
+    * @param name
+    *   of the actor
+    * @param initData
+    *   data to initialize the service
+    * @param context
+    *   the actor context of this actor system
+    * @param scheduler
+    *   the scheduler of simona
+    * @param extSimAdapter
+    *   the adapter for the external simulation
+    * @tparam T
+    *   type of [[ExtInputDataConnection]]
+    * @return
+    *   the reference to the service
+    */
+  private[setup] def setupInputService[T <: ExtInputDataConnection](
+      extInputDataConnection: T,
+      props: ClassicRef => Props,
+      name: String,
+      initData: T => InitializeServiceStateData,
+  )(implicit
+      context: ActorContext[_],
+      scheduler: ActorRef[SchedulerMessage],
+      extSimAdapter: ClassicRef,
+  ): ClassicRef = {
+
+    val extDataService = context.toClassic.simonaActorOf(
+      props(scheduler.toClassic),
+      name,
+    )
+
+    extDataService ! SimonaService.Create(
+      initData(extInputDataConnection),
+      ScheduleLock.singleKey(
+        context,
+        scheduler,
+        INIT_SIM_TICK,
+      ),
+    )
+
+    extInputDataConnection.setActorRefs(
+      extDataService,
+      extSimAdapter,
+    )
+
+    extDataService
+  }
+
+  /** Method for validating the external primary data connections.
+    * @param extPrimaryDataConnection
+    *   all external primary data connections
+    */
+  private[setup] def validatePrimaryData(
+      extPrimaryDataConnection: Set[ExtPrimaryDataConnection]
+  ): Unit = {
+    // check primary data for duplicate assets
+    val duplicateAssets: Iterable[UUID] =
+      extPrimaryDataConnection.toSeq
+        .flatMap(_.getPrimaryDataAssets.asScala)
+        .groupBy(identity)
+        .collect { case (uuid, Seq(_, _, _*)) => uuid }
```

Comment:
This looks like an unusual way of expressing this (that AI came up with?), I have never seen `_*` used in this way. Something more recognizable would, imho, be this:

```suggestion
        .collect { case (uuid, _ :: _ :: Nil) => uuid }
```

Or even easier, this:

```suggestion
        .collect { case (uuid, values) if values.size > 1 => uuid }
```